### PR TITLE
[MIG] migration business_requirement_deliverable from 10.0 to 12.0

### PR DESCRIPTION
**Pull request by ruter-lyu**
@elicoidal

### CODE REVIEW

**Comment by elicoidal** [comment]
Travis is complaining

File: business_requirement_deliverable/README.rst [hunk 1]
```diff
@@ -0,0 +1,189 @@
+.. figure:: https://img.shields.io/badge/licence-LGPL--3-blue.svg
```
Comment (elicoidal) [inline business_requirement_deliverable/README.rst]: svg->png

File: business_requirement_deliverable/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# © 2016-2019 Elico Corp (https://www.elico-corp.com).
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+{
+    'name': 'Business Requirement Deliverable',
+    'category': 'Business Requirements Management',
+    'summary': 'Manage the Business Requirement Deliverables and \
+                Resources for your customers',
+    'version': '10.0.1.0.0',
```
Comment (elicoidal) [inline business_requirement_deliverable/__manifest__.py]: 12

File: business_requirement_deliverable/data/br_report_paperformat.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
```
Comment (elicoidal) [inline business_requirement_deliverable/data/br_report_paperformat.xml:7]: copyright in all xml

File: business_requirement_deliverable/models/__init__.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+# © 2016-2019 Elico Corp (https://www.elico-corp.com).
```
Comment (elicoidal) [inline business_requirement_deliverable/models/__init__.py]: no need headers in init.py

File: business_requirement_deliverable/models/business.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+    @api.multi
+    @api.constrains('resource_type', 'uom_id')
+    def _check_description(self):
+        for resource in self:
+            if resource.resource_type == 'task' and (
+                    resource.uom_id.category_id != (
+                        self.env.ref('uom.uom_categ_wtime'))):
+                raise ValidationError(_(
+                    "When resource type is task, "
+                    "the uom category should be time"))
```
Comment (elicoidal) [inline business_requirement_deliverable/models/business.py]:
```suggestion
                    "the UoM category should be time"))
```

**Review by elicoidal — COMMENTED** [review]
not bad. Will need test

File: business_requirement_deliverable/models/business.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+                   ('confirmed', 'Confirmed'),
+                   ('approved', 'Approved'),
+                   ('stakeholder_approval', 'Stakeholder Approval'),
+                   ('in_progress', 'In progress'),
+                   ('done', 'Done'),
+                   ('cancel', 'Cancel'),
+                   ('drop', 'Drop'),
+                   ],
+        store=True,
+        readonly=True
```
Comment (nikul-serpentcs) [inline business_requirement_deliverable/models/business.py]: @ruter-lyu no need to `readonly=True` when field is related

File: business_requirement_deliverable/models/business.py [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+            uom_id = product.uom_id.id
+        if product.description_sale:
+            description += '\n' + product.description_sale
+        if not self.name:
+            self.name = description
+        if uom_id:
+            self.uom_id = uom_id
+
+    @api.onchange('resource_type')
+    def resource_type_change(self):
```
Comment (nikul-serpentcs) [inline business_requirement_deliverable/models/business.py]: no need to else
```
def resource_type_change(self):
        self.uom_id = self.env.ref('uom.product_uom_hour').id
        if self.resource_type == 'procurement':
            self.user_id = False
            self.uom_id = self.env.ref('uom.product_uom_unit').id


```

File: business_requirement_deliverable/models/business.py [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+                   ('confirmed', 'Confirmed'),
+                   ('approved', 'Approved'),
+                   ('stakeholder_approval', 'Stakeholder Approval'),
+                   ('in_progress', 'In progress'),
+                   ('done', 'Done'),
+                   ('cancel', 'Cancel'),
+                   ('drop', 'Drop'),
+                   ],
+        store=True,
+        readonly=True
```
Comment (nikul-serpentcs) [inline business_requirement_deliverable/models/business.py]: no need

File: business_requirement_deliverable/models/business.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+    )
+    uom_id = fields.Many2one(
+        comodel_name='uom.uom',
+        string='UoM',
+        required=True,
+        default=lambda self: self.env.ref('uom.product_uom_unit')
+    )
+    qty = fields.Float(
+        string='Quantity',
+        store=True,
```
Comment (nikul-serpentcs) [inline business_requirement_deliverable/models/business.py]: Does it need??

File: business_requirement_deliverable/models/business.py [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+    )
+    sale_price_unit = fields.Float(
+        string='Sales Price',
+        oldname='unit_price'
+    )
+    price_total = fields.Float(
+        compute='_compute_get_price_total',
+        string='Total Deliverable',
+        store=True,
+        readonly=True
```
Comment (nikul-serpentcs) [inline business_requirement_deliverable/models/business.py]: does it needs?

File: business_requirement_deliverable/models/business.py [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+    )
+    currency_id = fields.Many2one(
+        comodel_name='res.currency',
+        string='Currency',
+        readonly=True,
```
Comment (nikul-serpentcs) [inline business_requirement_deliverable/models/business.py]: `compute` always `readonly`

File: business_requirement_deliverable/security/business_requirement_deliverable_security.xml [hunk 12]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
Comment (nikul-serpentcs) [inline business_requirement_deliverable/security/business_requirement_deliverable_security.xml]: <odoo noupdate="1">

**Review by nikul-serpentcs — COMMENTED** [review]
Improved Code

File: business_requirement_deliverable/__manifest__.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    'website': 'https://github.com/OCA/business-requirement',
```
Comment (elicoidal) [inline business_requirement_deliverable/__manifest__.py:10]: Can you add the key `'maturity': 'stable'` in all modules?

**Review by elicoidal — CHANGES_REQUESTED** [review]

File: business_requirement_deliverable/models/business.py [hunk 14]
```diff
@@ -0,0 +1,445 @@
...
+                    "the UoM category should be time"))
+
+    @api.multi
+    def write(self, vals):
+        if vals.get('resource_type', '') == 'procurement':
+            vals['user_id'] = None
+        return super(BusinessRequirementResource, self).write(vals)
+
+
+class BusinessRequirementDeliverable(models.Model):
```
Comment (elicoidal) [inline business_requirement_deliverable/models/business.py]: Split the different models in different files

File: business_requirement_deliverable/models/business.py [hunk 15]
```diff
@@ -0,0 +1,445 @@
...
+    def _compute_get_currency(self):
+        for br in self:
+            if br.partner_id and br.pricelist_id.currency_id:
+                br.currency_id = br.pricelist_id.currency_id.id
+            else:
+                br.currency_id = self.env.user.company_id.currency_id.id
+
+    @api.multi
+    @api.onchange('partner_id')
+    def partner_id_change(self):
```
Comment (elicoidal) [inline business_requirement_deliverable/models/business.py:126]: It should be a simple warning.
We should add a onchange for the pricelist -> warning that the pricelist applies only to new deliverable lines.

File: business_requirement_deliverable/views/business_view.xml [hunk 16]
```diff
@@ -0,0 +1,346 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
Comment (elicoidal) [inline business_requirement_deliverable/views/business_view.xml:6]: copyright in all xml

File: business_requirement_deliverable/views/business_view.xml [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+                <br/>
+                <button class="oe_inline oe_stat_button"
+                    type="object" name="open_resource_line" icon="fa-tasks" >
+                    <field name="rl_count" string="Resource"
+                        widget="statinfo"
+                        modifiers="{'readonly': true}"/>
+                </button>
+            </xpath>
+            <field name="change_request" position="after">
+                <field name="pricelist_id" groups="business_requirement_deliverable.group_business_requirement_estimation"/>
```
Comment (elicoidal) [inline business_requirement_deliverable/views/business_view.xml]: improve readability (multiple)

File: business_requirement_deliverable/security/business_requirement_deliverable_security.xml [hunk 18]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+© 2016-2019 Elico Corp (https://www.elico-corp.com).
+License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+-->
+<odoo noupdate="1">
+
+       <record id="group_business_requirement_estimation" model="res.groups">
+           <field name="name">Business Requirement Estimation</field>
+           <field name="implied_ids" eval="[(4, ref('business_requirement.group_business_requirement_user'))]"/>
```
Comment (elicoidal) [inline business_requirement_deliverable/security/business_requirement_deliverable_security.xml]: improve readability

**Review by elicoidal — CHANGES_REQUESTED** [review]
not far. I will rework separately the fragments

**Comment by elicoidal** [comment]
not sure why coverall is not triggered.
~Travis is red~ Normal until first module is merged.

**Comment by elicoidal** [comment]
@JoJoJoJoJoJoJo @joycehuan please have a look

File: business_requirement_deliverable/models/business.py [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+            r.dl_count = len(r.deliverable_lines.ids)
+
+    @api.multi
+    def _compute_rl_count(self):
+        for r in self:
+            r.rl_count = len(r.resource_lines.ids)
+
+    @api.multi
+    def open_deliverable_line(self):
+        for self in self:
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement_deliverable/models/business.py:72]: Please improve this line.

File: business_requirement_deliverable/models/business.py [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+                    'tree_view_ref': 'business_requirement_deliverable.' +
+                    'view_business_requirement_deliverable_tree',
+                    'form_view_ref': 'business_requirement_deliverable.' +
+                    'view_business_requirement_deliverable_form',
+                    'default_business_requirement_id': br_id
+                }}
+
+    @api.multi
+    def open_resource_line(self):
+        for self in self:
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement_deliverable/models/business.py:95]: Same here.

File: business_requirement_deliverable/models/business_requirement_deliverable.py [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+        comodel_name='project.project',
+        related='business_requirement_id.project_id',
+        string='Project',
+        store=True
+    )
+
+    @api.multi
+    @api.onchange('business_requirement_id')
+    def business_requirement_id_change(self):
+        for deliverables in self:
```
Comment (JoJoJoJoJoJoJo) [inline business_requirement_deliverable/models/business_requirement_deliverable.py:82]: Improve the variable name.

**Review by JoJoJoJoJoJoJo — APPROVED** [review]
LGTM, some improvement.

**Comment by pedrobaeza** [comment]
Superseeded by #316

**Pull request closed without merging**